py_zerox/zerox/models/modellitellm.py

```diff
@@ -1,50 +1,82 @@
 import os
 import aiohttp
+import warnings
+import litellm
 from typing import List, Dict, Any, Optional
 
 # Package Imports
 from .base import BaseModel
 from .types import CompletionResponse
-from ..errors import MissingOpenAIAPIKeyException
+from ..errors import ModelAccessError, NotAVisionModel, MissingEnvironmentVariables
 from ..constants.messages import Messages
+from ..constants.prompts import Prompts
 from ..processor.image import encode_image_to_base64
 
+DEFAULT_SYSTEM_PROMPT = Prompts.DEFAULT_SYSTEM_PROMPT
 
-class OpenAI(BaseModel):
-    _instances: Dict[str, "OpenAI"] = {}
-    _system_prompt = """
-    Convert the following PDF page to markdown.
-    Return only the markdown with no explanation text.
-    Do not exclude any content from the page.
-    """
+
+class litellmmodel(BaseModel):
+    ## setting the default system prompt
+    _system_prompt = DEFAULT_SYSTEM_PROMPT
 
     def __init__(
         self,
-        api_key: Optional[str] = None,
+        model: Optional[str] = None,
+        **kwargs,
     ):
-        """Initializes the OpenAI model."""
-        if api_key is None:
-            api_key = self.get_api_key()
-
-        super().__init__(api_key)
-
-    @staticmethod
-    def get_api_key(
-        api_key: Optional[str] = None,
-    ) -> str:
-        """Gets the OpenAI API key from the environment variables incase it is not provided."""
-        if api_key is None:
-            api_key = os.getenv("OPENAI_API_KEY")
-        if api_key is None:
-            raise MissingOpenAIAPIKeyException()
-        return api_key
+        """
+        Initializes the Litellm model interface.
+        :param model: The model to use for generating completions, defaults to "gpt-4o-mini". Refer: https://docs.litellm.ai/docs/providers
+        :type model: str, optional
+        
+        :param kwargs: Additional keyword arguments to pass to self.completion -> litellm.completion. Refer: https://docs.litellm.ai/docs/providers and https://docs.litellm.ai/docs/completion/input
+        """
```

Comment:
```
        os.environ["OPENAI_API_KEY"] = kwargs.get("openai_api_key", '')
        os.environ["ANTHROPIC_API_KEY"] = kwargs.get("anthropic_api_key", '')
        os.environ["AZURE_API_KEY"] = kwargs.get("azure_api_key", '')
```
users should be able to set corresponding API keys via passing params into the zerox object.

Reply:
> ```
>         os.environ["OPENAI_API_KEY"] = kwargs.get("openai_api_key", '')
>         os.environ["ANTHROPIC_API_KEY"] = kwargs.get("anthropic_api_key", '')
>         os.environ["AZURE_API_KEY"] = kwargs.get("azure_api_key", '')
> ```
> 
> users should be able to set corresponding API keys via passing params into the zerox object.

@xdotli, I think as here also ultimately we'll be setting environment variables only so I would say better to use the environment variables instead of explicit params as this would just be consistent with litellm (incase more models are added to it). If you refer litellm docs, they also always prefer environment variables.

Reply:
@pradhyumna85 Hey Prad this can be valid, but sometimes users will not set the exact same env var name as we do. In this case exposing the options to the users will be very useful. Vercel's AI SDK is exposing to the users, pasted link here for reference:
https://sdk.vercel.ai/providers/ai-sdk-providers/openai#provider-instance

Reply:
@xdotli, may I resolve this conversation?

Reply:
> @pradhyumna85 Hey Prad this can be valid, but sometimes users will not set the exact same env var name as we do. In this case exposing the options to the users will be very useful. Vercel's AI SDK is exposing to the users, pasted link here for reference: https://sdk.vercel.ai/providers/ai-sdk-providers/openai#provider-instance

@xdotli, It looks like Vercel's sdk has different classes you need to import and use to call different models which is inherently different from litellm's api where you always interact with the **litellm.completion** api and I think that's the reason they also incline towards a [**environment vars** + **model name param**] approach to call model through a unified single api.

And for that, the [**litellmmodel.validate_environment**](https://github.com/pradhyumna85/zerox/blob/1adf7e272353118966760d4f45e44c1e2db57eff/py_zerox/zerox/models/modellitellm.py#L61) method in zerox, uses a **litellm utility** function **litellm.validate_environment** which exactly tells the - for a given model what all environment variables needs to be set and a boolean flag for if all needed env vars are set  and also if some environment variables are missing, it also return a list of the missing env vars name [Refer litellm.validate_environment](https://github.com/BerriAI/litellm/blob/de9a39e7c68ec677108f291fab8144fd91e3c30b/litellm/utils.py#L5234). We use this utility to raise helpful error message telling exactly what is wrong with the environment variables in [zerox's litellmmodel's **validate_environment** method.](https://github.com/pradhyumna85/zerox/blob/1adf7e272353118966760d4f45e44c1e2db57eff/py_zerox/zerox/models/modellitellm.py#L61).

Let me know if that made sense.